File: backend/routers/coach.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import JSONResponse, FileResponse
from fastapi.security import HTTPBearer
from sqlalchemy.orm import Session
from typing import List
from functools import wraps
from ..database import get_db
from .. import models


router = APIRouter(tags=["coach"])
# security = HTTPBearer()
# def get_coach_id(token: str = Depends(security), db: Session = Depends(get_db)):
def get_coach_id():
    """
    Placeholder function to get coach ID from session/token
    This will handle authentication and return the coach ID
    """
    # TODO: Implement actual token validation and coach ID extraction
    # For now, returning a placeholder coach ID
    coach_id = 1  # Placeholder
    
    # TODO: Add actual authentication logic here
    # - Validate token
    # - Extract user info from token
    # - Verify user is a coach
    # - Return coach ID or raise HTTPException if unauthorized
    
    return coach_id

@router.get("/student_search")
def search_students(search: str = "", coach_id: int = Depends(get_coach_id)):
    return [
        {"user_id": 1, "name": "John Doe", "email": "john@example"},
        {"user_id": 2, "name": "Jane Smith", "email": "jane@example"},
    ]

@router.get("/student_list")
def list_students(coach_id: int = Depends(get_coach_id)):
    return [
        {"user_id": 1, "name": "John Doe", "email": "john@example"},
        {"user_id": 2, "name": "Jane Smith", "email": "jane@example"},
    ]

@router.get("/student_video_list")
def list_student_videos(student_id: int, coach_id: int = Depends(get_coach_id)):
    return [
        {"video_id": 1, "video_status": "Completed"},
        {"video_id": 2, "video_status": "pending"},
    ]
@router.get("/student_video")
def student_video(student_id: int, video_id: int, coach_id: int = Depends(get_coach_id)):
    return FileResponse("./videos/movie.mp4", media_type="video/mp4")


@router.post("/comments/create")
def create_comment(coach_id: int = Depends(get_coach_id), student_id: str = "", video_id: str = "", timestamp_start: str = "", timestamp_end: str = "", comment: str = ""):
    return {
            "Success": True
            }

@router.get("/comments")
def get_comments(student_id: str, video_id: str, coach_id: int = Depends(get_coach_id)):
    return [
        {"comment_id": 1, "timestamp_start": "00:30", "timestamp_end": "00:45", "comment": "Great posture here"},
        {"comment_id": 2, "timestamp_start": "01:15", "timestamp_end": "01:20", "comment": "Work on frame"}
    ]

@router.delete("/comments/delete")
def delete_comment(coach_id: int = Depends(get_coach_id), student_id: str = "", video_id: str = "", timestamp_start: str = "", timestamp_end: str = "", comment: str = ""):
    return {
            "Success": True
            }

@router.post("/annotations/create")
def create_annotation(coach_id: int = Depends(get_coach_id), student_id: str = "", video_id: str = "", timestamp_start: str = "", timestamp_end: str = "", annotation: str = ""):
    return {
            "Success": True
            }

@router.get("/annotations")
def get_annotations(student_id: str, video_id: str, coach_id: int = Depends(get_coach_id)):
    return [
        {"annotation_id": 1, "timestamp": "00:15", "points": [{"x": 100, "y": 150}, {"x": 120, "y": 170}, {"x": 140, "y": 190}]},
        {"annotation_id": 2, "timestamp": "01:30", "points": [{"x": 250, "y": 300}, {"x": 270, "y": 320}]}
    ]

@router.delete("/annotations/delete")
def delete_annotation(coach_id: int = Depends(get_coach_id), student_id: str = "", video_id: str = "", timestamp_start: str = "", timestamp_end: str = "", annotation: str = ""):
    return {
            "Success": True
            }

File: frontend/src/services/coach.js
const API_BASE_URL = process.env.REACT_APP_API_BASE_URL || 'http://localhost:8000';
const BASE_URL = `${API_BASE_URL}/coach`;

export async function searchStudents(search = '') {
    const response = await fetch(`${BASE_URL}/student_search?search=${encodeURIComponent(search)}`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function listStudents() {
    const response = await fetch(`${BASE_URL}/student_list`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function listStudentVideos(studentId) {
    const response = await fetch(`${BASE_URL}/student_video_list?student_id=${studentId}`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function getStudentVideo(studentId, videoId) {
    const response = await fetch(`${BASE_URL}/student_video?student_id=${studentId}&video_id=${videoId}`, {
        method: 'GET',
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.blob();
}

export async function createComment(studentId, videoId, timestampStart, timestampEnd, comment) {
    const response = await fetch(`${BASE_URL}/comments/create`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            student_id: studentId,
            video_id: videoId,
            timestamp_start: timestampStart,
            timestamp_end: timestampEnd,
            comment: comment,
        }),
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function getComments(studentId, videoId) {
    const response = await fetch(`${BASE_URL}/comments?student_id=${studentId}&video_id=${videoId}`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function deleteComment(studentId, videoId, timestampStart, timestampEnd, comment) {
    const response = await fetch(`${BASE_URL}/comments/delete`, {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            student_id: studentId,
            video_id: videoId,
            timestamp_start: timestampStart,
            timestamp_end: timestampEnd,
            comment: comment,
        }),
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function createAnnotation(studentId, videoId, timestampStart, timestampEnd, annotation) {
    const response = await fetch(`${BASE_URL}/annotations/create`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            student_id: studentId,
            video_id: videoId,
            timestamp_start: timestampStart,
            timestamp_end: timestampEnd,
            annotation: annotation,
        }),
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function getAnnotations(studentId, videoId) {
    const response = await fetch(`${BASE_URL}/annotations?student_id=${studentId}&video_id=${videoId}`, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}

export async function deleteAnnotation(studentId, videoId, timestampStart, timestampEnd, annotation) {
    const response = await fetch(`${BASE_URL}/annotations/delete`, {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            student_id: studentId,
            video_id: videoId,
            timestamp_start: timestampStart,
            timestamp_end: timestampEnd,
            annotation: annotation,
        }),
    });
    
    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return response.json();
}
